pageserver/src/metrics.rs

```diff
@@ -0,0 +1,418 @@
+            .get_metric_with_label_values(&["init logical size", &tenant_id, &timeline_id])
+            .unwrap();
+        let load_layer_map_histo = STORAGE_TIME
+            .get_metric_with_label_values(&["load layer map", &tenant_id, &timeline_id])
+            .unwrap();
+        let last_record_gauge = LAST_RECORD_LSN
+            .get_metric_with_label_values(&[&tenant_id, &timeline_id])
+            .unwrap();
+        let wait_lsn_time_histo = WAIT_LSN_TIME
+            .get_metric_with_label_values(&[&tenant_id, &timeline_id])
+            .unwrap();
+        let current_physical_size_gauge = CURRENT_PHYSICAL_SIZE
+            .get_metric_with_label_values(&[&tenant_id, &timeline_id])
+            .unwrap();
+        let current_logical_size_gauge = CURRENT_LOGICAL_SIZE
+            .get_metric_with_label_values(&[&tenant_id, &timeline_id])
+            .unwrap();
+        let num_persistent_files_created = NUM_PERSISTENT_FILES_CREATED.clone();
+        let persistent_bytes_written = PERSISTENT_BYTES_WRITTEN.clone();
+
+        TimelineMetrics {
+            tenant_id,
+            timeline_id,
+            reconstruct_time_histo,
+            materialized_page_cache_hit_counter,
+            flush_time_histo,
+            compact_time_histo,
+            create_images_time_histo,
+            init_logical_size_histo,
+            load_layer_map_histo,
+            last_record_gauge,
+            wait_lsn_time_histo,
+            current_physical_size_gauge,
+            current_logical_size_gauge,
+            num_persistent_files_created,
+            persistent_bytes_written,
+        }
+    }
+}
+
+impl Drop for TimelineMetrics {
+    fn drop(&mut self) {
+        let tenant_id = &self.tenant_id;
+        let timeline_id = &self.timeline_id;
+        let _ = RECONSTRUCT_TIME.remove_label_values(&[tenant_id, timeline_id]);
+        let _ = MATERIALIZED_PAGE_CACHE_HIT.remove_label_values(&[tenant_id, timeline_id]);
+        let _ = LAST_RECORD_LSN.remove_label_values(&[tenant_id, timeline_id]);
+        let _ = WAIT_LSN_TIME.remove_label_values(&[tenant_id, timeline_id]);
+        let _ = CURRENT_PHYSICAL_SIZE.remove_label_values(&[tenant_id, timeline_id]);
+        let _ = CURRENT_LOGICAL_SIZE.remove_label_values(&[tenant_id, timeline_id]);
+
+        for op in STORAGE_TIME_OPERATIONS {
+            let _ = STORAGE_TIME.remove_label_values(&[op, tenant_id, timeline_id]);
+        }
+        for op in STORAGE_IO_TIME_OPERATIONS {
+            let _ = STORAGE_IO_TIME.remove_label_values(&[op, tenant_id, timeline_id]);
+        }
+
+        for op in STORAGE_IO_SIZE_OPERATIONS {
+            let _ = STORAGE_IO_SIZE.remove_label_values(&[op, tenant_id, timeline_id]);
+        }
+
+        for op in SMGR_QUERY_TIME_OPERATIONS {
+            let _ = SMGR_QUERY_TIME.remove_label_values(&[op, tenant_id, timeline_id]);
+        }
+    }
+}
+
+pub fn remove_tenant_metrics(tenant_id: &str) {
```

Comment:
Hmm, might have misunderstood, but I thought `Repository` wouldn’t be tenant specific, so that wouldn’t drop the metrics when tenant is detached 🤔

I’ll take another look.

Reply:
Repository is tenant specific. Tenants cannot share the same repo

See Arthur's [comment](https://github.com/neondatabase/neon/pull/2328#pullrequestreview-1094760288) we'll probably want to remove metrics for inactive tenants/timelines so may worth consider alternatives for using Drop. But thats a good start anyway :)

Reply:
I guess using drop trait indeed makes sense now over calling cleanup from `tenant_mgr`. Need to familiarize myself a bit more with the codebase and take a second round with what Arthur suggested. At least comparing to safekeeper, it seems the strategy is built around metrics collector rather than removing label values like I'm trying to do here, so a few things to consider.